## Goals of the study:  
Generate a value proposition for a new product,<br>
In order to develop new apps for book lovers.

### Table of Contents:
* [Setup.](#Setup)
* [Functions.](#Functions)  
* [Study the tables](#Study)
* [1. Find the number of books released after January 1, 2000.](#1.1)
* [2. Find the number of user reviews and the average rating for each book.](#2)
* [3. Identify the publisher that has released the greatest number of books](#3)
* [4. Identify the author with the highest average book rating.](#4)
* [1. Find the number of books released after January 1, 2000.](#1)
* [1. Find the number of books released after January 1, 2000.](#1)




### Setup.<a class="anchor" id="Setup"></a>

In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'practicum_student',         # username
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Functions.<a class="anchor" id="Functions"></a>

In [2]:
def print_table_head(table):
    '''
    print a tables with 5 rows
    '''
    print()
    print('************ ', table, ' ************')
    query = "SELECT * FROM " + table + " LIMIT 5"
    display(pd.io.sql.read_sql(query, con = engine))

### Study the tables (print the first rows).<a class="anchor" id="Study"></a>

create a list of all the tables

In [3]:
tables = ['BOOKS', 'authors', 'publishers', 'ratings', 'reviews']

print all of the tables with a few lines

In [4]:
for table in tables:
    print_table_head(table)


************  BOOKS  ************


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268


************  authors  ************


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd


************  publishers  ************


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company


************  ratings  ************


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2


************  reviews  ************


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

### 1. Find the number of books released after January 1, 2000.<a class="anchor" id="1"></a>

In [5]:
query ='''
SELECT COUNT(book_id) AS books_amount
FROM BOOKS
WHERE publication_date > '2000-01-01';
'''
#print(query)

display(pd.io.sql.read_sql(query, con = engine))

books_amount
0           819

conclusion: there were 819 books that were released after January 1, 2000.

### 2. Find the number of user reviews and the average rating for each book.<a class="anchor" id="2"></a>

In [6]:
query ='''
SELECT 
    books.book_id,
    books.title,
    AVG(rating) AS rating_avg,
    count (distinct  reviews.review_id) AS reviews_CNT
FROM books  
INNER JOIN ratings ON books.book_id = ratings.book_id
INNER JOIN reviews ON books.book_id = reviews.book_id
GROUP BY
    books.book_id    
ORDER BY
    reviews_CNT DESC  
'''
#print(query)

display(pd.io.sql.read_sql(query, con = engine))

book_id                                title  rating_avg  reviews_cnt
0        948              Twilight (Twilight  #1)    3.662500            7
1        854                             The Road    3.772727            6
2        656                       The Book Thief    4.264151            6
3        734                     The Glass Castle    4.206897            6
4        963                  Water for Elephants    3.977273            6
..       ...                                  ...         ...          ...
989      465    Naked Empire (Sword of Truth  #8)    3.500000            1
990      446                  Moo  Baa  La La La!    3.000000            1
991      431  Merrick (The Vampire Chronicles #7)    4.000000            1
992       92                            Babyville    3.500000            1
993      413                           Lysistrata    4.000000            1

[994 rows x 4 columns]

some aditional check of the data

Lets look at books with more then 3 reviews

In [7]:
query ='''
SELECT 
    books.book_id,
    books.title,
    AVG(rating) AS rating_avg,
    count (distinct  reviews.review_id) AS reviews_CNT
FROM books  
INNER JOIN ratings ON books.book_id = ratings.book_id
INNER JOIN reviews ON books.book_id = reviews.book_id 
GROUP BY
    books.book_id 
HAVING     
    count (distinct  reviews.review_id) > 3
ORDER BY
    reviews_CNT DESC  
LIMIT 50    
'''
#print(query)

display(pd.io.sql.read_sql(query, con = engine))

book_id                                              title  rating_avg  \
0       948                            Twilight (Twilight  #1)    3.662500   
1       963                                Water for Elephants    3.977273   
2       779  The Lightning Thief (Percy Jackson and the Oly...    4.080645   
3       627                                      The Alchemist    3.789474   
4       733                          The Giver (The Giver  #1)    3.750000   
5       207                                    Eat  Pray  Love    3.395833   
6       302  Harry Potter and the Prisoner of Azkaban (Harr...    4.414634   
7       673                             The Catcher in the Rye    3.825581   
8       734                                   The Glass Castle    4.206897   
9       299  Harry Potter and the Chamber of Secrets (Harry...    4.287500   
10      497                          Outlander (Outlander  #1)    4.125000   
11      695  The Curious Incident of the Dog in the Night-Time    4.081081   
12      854                                           The Road    3.772727   
13      656                                     The Book Thief    4.264151   
14      696             The Da Vinci Code (Robert Langdon  #2)    3.830508   
15      750                The Hobbit  or There and Back Again    4.125000   
16       51                           A Tree Grows in Brooklyn    4.250000   
17       90                                     Atlas Shrugged    3.583333   
18      577                                Slaughterhouse-Five    4.028571   
19       84    Anne of Green Gables (Anne of Green Gables  #1)    4.000000   
20       79                                        Animal Farm    3.729730   
21      471                                    Never Let Me Go    3.928571   
22      481                                     Norwegian Wood    3.888889   
23      478                                   Nineteen Minutes    4.000000   
24      545                                   Romeo and Juliet    3.787879   
25      488                                    Of Mice and Men    3.622951   
26      420  Marley and Me: Life and Love With the World's ...    3.800000   
27       97                                          Bel Canto    3.375000   
28      419                        Marked (House of Night  #1)    3.571429   
29      581                     Snow Flower and the Secret Fan    4.636364   
30      426                                Memoirs of a Geisha    4.195652   
31      405                                  Lord of the Flies    3.901408   
32      399                                       Little Women    4.192308   
33      408                        Love in the Time of Cholera    4.083333   
34      437                                          Middlesex    3.421053   
35      162                                           Coraline    4.133333   
36      304                         Hatchet (Brian's Saga  #1)    4.200000   
37      300  Harry Potter and the Half-Blood Prince (Harry ...    4.246575   
38      364                                          Jane Eyre    4.102041   
39      374                                 Kafka on the Shore    4.125000   
40       70                  American Gods (American Gods  #1)    4.000000   
41      194                                            Dracula    3.296296   
42       74                           And Then There Were None    4.454545   
43      417                           Man's Search for Meaning    4.000000   
44      145                                    Charlotte's Web    3.956522   
45       75               Angels & Demons (Robert Langdon  #1)    3.678571   
46      301  Harry Potter and the Order of the Phoenix (Har...    4.186667   
47      232                Extremely Loud and Incredibly Close    3.416667   
48      472                      Neverwhere (London Below  #1)    4.083333   
49      320                                  Holes (Holes  #1)    3.967742   

    reviews_cnt  
0      

books with low rating, below 3

In [8]:
query ='''
SELECT 
    books.book_id,
    books.title,
    AVG(rating) AS rating_avg,
    count (distinct  reviews.review_id) AS reviews_CNT
FROM books  
INNER JOIN ratings ON books.book_id = ratings.book_id
INNER JOIN reviews ON books.book_id = reviews.book_id 
GROUP BY
    books.book_id 
HAVING     
    AVG(rating) < 3
ORDER BY
    reviews_CNT DESC  
LIMIT 50    
'''
#print(query)

display(pd.io.sql.read_sql(query, con = engine))

book_id                                              title  rating_avg  \
0       202                                      Drowning Ruth    2.000000   
1       552  Saving the World and Other Extreme Sports (Max...    2.666667   
2       557                      Seven Up (Stephanie Plum  #7)    2.750000   
3       272               Furies of Calderon (Codex Alera  #1)    2.666667   
4       765                          The Jane Austen Book Club    2.666667   
5       772                             The Kitchen God's Wife    2.333333   
6       794                                  The Mermaid Chair    2.333333   
7        96  Beauty: A Retelling of the Story of Beauty and...    2.666667   
8       915  The World Is Flat: A Brief History of the Twen...    2.250000   
9       200                                       Dreamcatcher    2.800000   
10      412                   Lucy Sullivan Is Getting Married    2.500000   
11      445                                            Monster    2.500000   
12      452                                          Mr. Maybe    2.750000   
13      470          Neither Here nor There: Travels in Europe    2.500000   
14      493                                         Open House    2.500000   
15      525                                     Prep : A Novel    2.500000   
16      747         The High Lord (Black Magician Trilogy  #3)    2.500000   
17      822                                     The Penelopiad    2.500000   
18      858                            The Rules of Attraction    2.500000   
19      911                       The Winter of Our Discontent    2.500000   
20      983  Wild at Heart: Discovering the Secret of a Man...    2.500000   
21       72                                          Amsterdam    2.500000   
22      164                            Count Zero (Sprawl  #2)    2.500000   
23      193                                        Dr. Faustus    2.500000   
24      278                                        Ghost Story    2.666667   
25      288                      Gossip Girl (Gossip Girl  #1)    2.500000   
26      303                               Harvesting the Heart    1.500000   
27      313              Heretics of Dune (Dune Chronicles #5)    2.500000   
28      316                  His Excellency: George Washington    2.000000   
29      371                                              Junky    2.000000   
30      929                                      Thunderstruck    2.500000   
31      939                                   Tropic of Cancer    2.500000   
32        2                 1 000 Places to See Before You Die    2.500000   
33      318                                  History of Beauty    2.500000   
34      943               Tsubasa: RESERVoir CHRoNiCLE  Vol. 1    2.500000   

    reviews_cnt  
0             3  
1             3  
2             3  
3             3  
4             3  
5             3  
6             3  
7             3  
8             3  
9             3  
10            2  
11            2  
12            2  
13            2  
14            2  
15            2  
16            2  
17            2  
18            2  
19            2  
20            2  
21            2  
22            2  
23            2  
24            2  
25            2  
26            2  
27            2  
28            2  
29            2  
30            2  
31            2  
32            1  
33            1  
34            1

Conclusions:  <br>
6 reviews are considered to be a large number.<br>
Most books that receive 5 or 6 review tend to also have an average high rating.<br>
Books with few reviews, less than 3, tend to have a low rating.


### 3. Identify the publisher that has released the greatest number of books with more than 50 pages.<a class="anchor" id="3"></a>

In [9]:
query ='''
SELECT 
    publishers.publisher,
    books.publisher_id, 
    count(books.publisher_id) AS published_cnt
FROM 
    books
    INNER JOIN publishers ON publishers.publisher_id = books.publisher_id
WHERE
    num_pages > 50
GROUP BY 
    books.publisher_id,
    publishers.publisher
ORDER BY   
    published_cnt DESC
LIMIT 1    
'''
#print(query)

display(pd.io.sql.read_sql(query, con = engine))

publisher  publisher_id  published_cnt
0  Penguin Books           212             42

Conclusions: <br> 
Penguin Books is the publisher that has published the most books with more then 50 pages.

### 4. Identify the author with the highest average book rating .<a class="anchor" id="4"></a> <br>
look only at books with at least 50 ratings

In [10]:
query ='''
SELECT
    books.author_id, 
    books.title,
    authors.author,
    count(ratings.rating_id) AS rate_cnt,
    AVG(ratings.rating) AS rate
FROM 
    books
    INNER JOIN ratings ON ratings.book_id = books.book_id
    INNER JOIN authors ON authors.author_id = books.author_id
    
GROUP BY
    books.author_id,
    books.title,
    authors.author
HAVING     
    count(ratings.rating_id) > 50
ORDER BY
    rate DESC
LIMIT 1    
'''
#print(query)

display(pd.io.sql.read_sql(query, con = engine))

author_id                                              title  \
0        236  Harry Potter and the Prisoner of Azkaban (Harr...   

                       author  rate_cnt      rate  
0  J.K. Rowling/Mary GrandPré        82  4.414634

Conclusions: <br> 
the author with the highest book rating are actually a team J.K. Rowling & Mary GrandPré .<br> 
For the book: Harry Potter and the Prisoner of Azkaban


### 5. Find the average number of text reviews among users who rated more than 50 books.<a class="anchor" id="5"></a>

did any user rated the same book twice ?

In [11]:
query ='''
SELECT
   COUNT(book_id) AS all,
   COUNT(DISTINCT book_id) AS distinc,
   (COUNT(book_id) - COUNT(DISTINCT book_id)) AS dif
FROM 
    ratings      
GROUP BY
    username
HAVING   
    COUNT(book_id) - COUNT(DISTINCT book_id) > 0
'''
print(query)


display(pd.io.sql.read_sql(query, con = engine))


SELECT
   COUNT(book_id) AS all,
   COUNT(DISTINCT book_id) AS distinc,
   (COUNT(book_id) - COUNT(DISTINCT book_id)) AS dif
FROM 
    ratings      
GROUP BY
    username
HAVING   
    COUNT(book_id) - COUNT(DISTINCT book_id) > 0



Empty DataFrame
Columns: [all, distinc, dif]
Index: []

no user has rated the same book twice

In [12]:
query ='''
SELECT       
    AVG(rev_cnt)
FROM 
    (SELECT
       DISTINCT ratings.username,
       COUNT  (DISTINCT reviews.review_id) AS rev_cnt
    FROM 
        ratings   
        INNER JOIN reviews ON reviews.username = ratings.username    
    GROUP BY
        ratings.username
    HAVING 
        COUNT(ratings.book_id) > 50) AS sub_rat    
'''
#print(query)

#display(pd.io.sql.read_sql(query, con = engine))

Conclusions: <br>
the average number or text reviews for readers who have rated more then 50 books is 17.45 <br>
This mean that they only review about every 3rd book they rated.


create a list with users with more the 50 ratings

In [13]:
query ='''
SELECT       
   username,
   count(book_id) AS rat_cnt
FROM    
    ratings  
GROUP BY
    username
HAVING 
    count(book_id) > 50
'''
#print(query)

display(pd.io.sql.read_sql(query, con = engine))

username  rat_cnt
0     sfitzgerald       55
1  jennifermiller       53
2          xdavis       51
3          paul88       56
4      martinadam       56
5       richard89       55

for that list of users find out haw many reviews they have

In [14]:
query ='''
SELECT
    reviews.username,
    count(reviews.book_id) AS rev_cnt
FROM
    (SELECT       
       username,
       count(book_id) AS rat_cnt
    FROM    
        ratings  
    GROUP BY
        username
    HAVING 
        count(book_id) > 50) AS user_list
    INNER JOIN reviews on user_list.username = reviews.username
WHERE
    user_list.username = reviews.username
GROUP BY
    reviews.username
   
'''
#print(query)

display(pd.io.sql.read_sql(query, con = engine))

username  rev_cnt
0     sfitzgerald       28
1  jennifermiller       25
2          xdavis       18
3          paul88       22
4      martinadam       27
5       richard89       26

find the average

In [15]:
query ='''
SELECT 
    AVG(rev_cnt)
FROM (
    SELECT 
        count(reviews.book_id) AS REV_cnt
    FROM
        (SELECT       
           username,
           count(book_id) AS rat_cnt
        FROM    
            ratings  
        GROUP BY
            username
        HAVING 
            count(book_id) > 50) AS user_list
        INNER JOIN reviews on user_list.username = reviews.username
    WHERE
        user_list.username = reviews.username
    GROUP BY
        reviews.username) AS rev_top

'''
#print(query)

display(pd.io.sql.read_sql(query, con = engine))

avg
0  24.333333

Conclusions: <br>
the average number of text reviews for readers who have rated more then 50 books is 24.3 <br>
This mean that they only review about every 2nd book they rated.